src/scanner.c

```diff
@@ -0,0 +1,403 @@
+  }
+
+  // we're in the body of the raw string; start looping until
+  // we find the matching closing bracket
+  for (; lexer->lookahead != 0; lexer->advance(lexer, false)) {
+
+    // consume a closing bracket
+    if (lexer->lookahead != closing_bracket) {
+      continue;
+    }
+    lexer->advance(lexer, false);
+
+    // consume hyphens
+    bool hyphens_ok = true;
+    for (int i = 0; i < hyphen_count; i++) {
+      if (lexer->lookahead != '-') {
+        hyphens_ok = false;
+        break;
+      }
+      lexer->advance(lexer, false);
+    }
+
+    if (!hyphens_ok) {
+      continue;
+    }
+
+    // consume a closing quote character
+    if (lexer->lookahead != quote) {
+      continue;
+    }
+    lexer->advance(lexer, false);
+
+    // success!
+    lexer->mark_end(lexer);
+    lexer->result_symbol = RAW_STRING_LITERAL;
+    return true;
+
+  }
+
+  // if we get here, this implies we hit eof (and so we have
+  // an unclosed raw string)
+  return false;
+
+}
+
+static bool scan(TSLexer* lexer, Stack* stack, const bool* valid_symbols) {
+  consume_whitespace_and_ignored_newlines(lexer, stack);
+
+  // check for semi-colons
+  if (valid_symbols[SEMICOLON] && lexer->lookahead == ';') {
+    lexer->advance(lexer, false);
+    lexer->mark_end(lexer);
+    lexer->result_symbol = SEMICOLON;
+    return true;
+  }
+
+  // check for an open bracket
+  if (valid_symbols[OPEN_PAREN] && lexer->lookahead == '(') {
+    lexer->advance(lexer, false);
+    lexer->mark_end(lexer);
+    lexer->result_symbol = OPEN_PAREN;
+    stack_push(stack, SCOPE_PAREN);
+    return true;
+  }
+
+  if (valid_symbols[CLOSE_PAREN] && lexer->lookahead == ')') {
+    lexer->advance(lexer, false);
+    lexer->mark_end(lexer);
+    lexer->result_symbol = CLOSE_PAREN;
+    stack_pop(stack, SCOPE_PAREN);
+    return true;
+  }
+
+  if (valid_symbols[OPEN_BRACE] && lexer->lookahead == '{') {
+    lexer->advance(lexer, false);
+    lexer->mark_end(lexer);
+    lexer->result_symbol = OPEN_BRACE;
+    stack_push(stack, SCOPE_BRACE);
+    return true;
+  }
+
+  if (valid_symbols[CLOSE_BRACE] && lexer->lookahead == '}') {
+    lexer->advance(lexer, false);
+    lexer->mark_end(lexer);
+    lexer->result_symbol = CLOSE_BRACE;
+    stack_pop(stack, SCOPE_BRACE);
+    return true;
+  }
+
+  if (valid_symbols[OPEN_BRACKET] || valid_symbols[OPEN_BRACKET2]) {
+    if (lexer->lookahead == '[') {
+      lexer->advance(lexer, false);
+      if (lexer->lookahead == '[') {
+        lexer->advance(lexer, false);
+        lexer->mark_end(lexer);
+        lexer->result_symbol = OPEN_BRACKET2;
+        stack_push(stack, SCOPE_BRACKET2);
+      } else {
+        lexer->mark_end(lexer);
+        lexer->result_symbol = OPEN_BRACKET;
+        stack_push(stack, SCOPE_BRACKET);
+      }
+      return true;
+    }
+  }
+
+  if (valid_symbols[CLOSE_BRACKET] || valid_symbols[CLOSE_BRACKET2]) {
+    if (lexer->lookahead == ']') {
+      lexer->advance(lexer, false);
+      if (lexer->lookahead == ']') {
```

Comment:
There are no changes in the underlying `scan()` function beyond switching out push/pop for the new stack variants except for this line where we previously had

```
if (lexer->lookahead == ']' && peek() == OPEN_BRACKET2) {
```

If the 2nd half of this was `FALSE` then somehow we had an upcoming `]]` but weren't in a `[[` scope.

Previously if that was the case then we dropped into the `else {` condition where we marked the result symbol as `CLOSE_BRACKET`? And then tried to pop a `]` off the stack? That seems like it would not really work well as an error recovery strategy, so I've removed the `peek() == OPEN_BRACKET2` of this check.

Reply:
I think I've handled this pretty elegantly now in https://github.com/r-lib/tree-sitter-r/pull/75/commits/df41984ba78dee164b63b64ce25562a1dc779587